File: Codes/InsertionSortList.cpp
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        if (!head)
        	return NULL;
        ListNode *n0;
        n0 = new ListNode(0);
        n0->next = head;
        ListNode *nICur = head->next, *nIPrev = head, *nJCur = head, *nJPrev = n0;
        
        while (nICur) {
            nJPrev = n0;
            nJCur = n0->next;
            while (nICur->val > nJCur->val && nICur != nJCur) {
                nJPrev = nJCur;
                nJCur = nJCur->next;
            }
            nIPrev->next = nICur->next;
            nJPrev->next = nICur;
            if (nICur != nJCur) {
                nICur->next = nJCur;
                nICur = nIPrev->next;
            } else {
                nIPrev = nICur;
                nICur = nICur->next;
            }
        }
        
        head = n0->next;
        delete n0;
        return head;
    }
};

File: Codes/InsertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None:
            return head
        h0 = ListNode(0)
        h0.next = head
        curr, prev = head.next, head
        while curr != None:
            nd, ndPrev = h0.next, h0
            while nd != curr:
                if curr.val < nd.val:
                    #insert curr before nd
                    #curr.next, prev.next, ndPrev.next are affected
                    prev.next = curr.next
                    curr.next = nd
                    ndPrev.next = curr
                nd, ndPrev = nd.next, ndPrev.next
            curr, prev = curr.next, prev.next
        return h0.next
